src/libostree/ostree-mutable-tree.c

```diff
@@ -153,6 +173,83 @@ invalidate_contents_checksum (OstreeMutableTree *self)
   }
 }
 
+/* Go from state LAZY to state WHOLE by reading the tree from disk */
+static gboolean
+_ostree_mutable_tree_make_whole (OstreeMutableTree           *self,
+                                 GCancellable                *cancellable,
+                                 GError                     **error)
+{
+  if (self->state == MTREE_STATE_WHOLE)
+    return TRUE;
+
+  g_assert_cmpuint (self->state, ==, MTREE_STATE_LAZY);
+  g_assert_nonnull (self->repo);
+  g_assert_nonnull (self->contents_checksum);
+  g_assert_nonnull (self->metadata_checksum);
+  g_assert_cmpuint (g_hash_table_size (self->files), ==, 0);
+  g_assert_cmpuint (g_hash_table_size (self->subdirs), ==, 0);
+
+  g_autoptr(GVariant) dirtree = NULL;
+  if (!ostree_repo_load_variant (self->repo, OSTREE_OBJECT_TYPE_DIR_TREE,
+                                 self->contents_checksum, &dirtree, error))
+    return FALSE;
+
+  {
+    g_autoptr(GVariant) dir_file_contents = g_variant_get_child_value (dirtree, 0);
+    GVariantIter viter;
+    g_variant_iter_init (&viter, dir_file_contents);
+    const char *fname;
+    GVariant *contents_csum_v = NULL;
+    while (g_variant_iter_loop (&viter, "(&s@ay)", &fname, &contents_csum_v))
+      {
+        char tmp_checksum[OSTREE_SHA256_STRING_LEN + 1];
+        _ostree_checksum_inplace_from_bytes_v (contents_csum_v, tmp_checksum);
+        g_hash_table_insert (self->files, g_strdup (fname),
+            g_strdup (tmp_checksum));
+      }
+  }
+
+  /* Process subdirectories */
+  {
+    g_autoptr(GVariant) dir_subdirs = g_variant_get_child_value (dirtree, 1);
+    const char *dname;
+    GVariant *subdirtree_csum_v = NULL;
+    GVariant *subdirmeta_csum_v = NULL;
+    GVariantIter viter;
+    g_variant_iter_init (&viter, dir_subdirs);
+    while (g_variant_iter_loop (&viter, "(&s@ay@ay)", &dname,
+                                &subdirtree_csum_v, &subdirmeta_csum_v))
+      {
+        char subdirtree_checksum[OSTREE_SHA256_STRING_LEN+1];
+        _ostree_checksum_inplace_from_bytes_v (subdirtree_csum_v, subdirtree_checksum);
+        char subdirmeta_checksum[OSTREE_SHA256_STRING_LEN+1];
+        _ostree_checksum_inplace_from_bytes_v (subdirmeta_csum_v, subdirmeta_checksum);
+        insert_child_mtree (self, dname, ostree_mutable_tree_new_from_checksum (
+            self->repo, subdirtree_checksum, subdirmeta_checksum));
+      }
+  }
+
+  g_clear_object (&self->repo);
+  self->state = MTREE_STATE_WHOLE;
+  return TRUE;
+}
+
+/* _ostree_mutable_tree_make_whole can fail if state == MTREE_STATE_LAZY, but
+ * we have getters that preceed the existence of MTREE_STATE_LAZY which can't
+ * return errors.  So instead this function will fail and print a warning. */
```

Comment:
Hmm.  One approach I've taken in other places for things like this is to cache the error, changing the object into an "error state".  Have all API functions be "noops", and return the error at the next opportunity.

We might also be able to add a new API to check for and consume that error? 

Not a hard requirement right now though.